File: src/game/xp.js
/* Experience gems: tiered by value, pooled, and merged into an existing gem
 * when the field saturates so a late swarm never loses experience to the cap. */
'use strict';
(function (WS) {

  const XP = { pool: null, vacuumTimer: 0 };

  XP.init = function () {
    this.pool = new WS.Pool(() => ({}), null, WS.CONST.MAX_GEMS);
    this.vacuumTimer = 0;
  };

  XP.clear = function () {
    this.vacuumTimer = 0;
    if (this.pool) this.pool.releaseAll();
  };

  function style(gem) {
    const C = WS.CONST.COLORS;
    // Purely how big it looks. What it takes to collect one is gem.radius,
    // which is a flat 8 for every tier and is deliberately not tied to this -
    // a gem must not become easier to pick up by being worth more.
    if (gem.value >= 25) { gem.size = 9.45; gem.colour = C.gemHigh; gem.tier = 2; }
    else if (gem.value >= 8) { gem.size = 7.88; gem.colour = C.gemMid; gem.tier = 1; }
    else { gem.size = 6.3; gem.colour = C.gemLow; gem.tier = 0; }
  }

  /* Gems merge on contact rather than piling up.
   *
   * A four-minute run used to leave 200-odd gems lying on the field, and two
   * hundred 6px marks scattered over the ground do not read as loot - they
   * read as static. Merging on spawn keeps the field in the tens: fewer, and
   * each one worth more, which is also a better thing to walk toward. No
   * experience is lost either way, since the merge target keeps the value.
   *
   * MERGE is deliberately smaller than the pickup radius, so gems only fuse
   * when they were going to overlap anyway. */
  const MERGE = 26, MERGE2 = MERGE * MERGE;

  XP.spawnGem = function (x, y, value) {
    if (value <= 0) return;
    const gx = x + WS.randRange(-6, 6), gy = y + WS.randRange(-6, 6);

    // Fold into a neighbour if there is one. The scan is linear, but it ends
    // on the first hit and the field is densest exactly when hits are likely.
    for (let i = 0; i < this.pool.count; i++) {
      const g = this.pool.active[i];
      const dx = g.x - gx, dy = g.y - gy;
      if (dx * dx + dy * dy < MERGE2) {
        g.value += value;
        style(g);
        g.pop = 0.18;                       // a small swell, so a merge reads
        return;
      }
    }

    // The field is full and nothing was close enough: fold into a random gem
    // rather than dropping the experience on the floor.
    if (this.pool.count >= WS.CONST.MAX_GEMS) {
      const gem = this.pool.active[WS.randInt(0, this.pool.count - 1)];
      gem.value += value;
      style(gem);
      gem.pop = 0.18;
      return;
    }
    const gem = this.pool.acquire();
    if (!gem) return;
    gem.x = gx;
    gem.y = gy;
    gem.value = value;
    gem.radius = 8;
    gem.pop = 0.18;
    // A fixed orientation, set once. Gems used to rotate at 2 rad/s, which on
    // a 6px diamond is not rotation - it is the shape changing width every
    // frame, and across a field of them it reads as flicker.
    gem.spin = WS.random() * WS.TAU;
    style(gem);
  };

  /** Lodestone: a brief one-shot sweep of everything currently on the field.
   *  Timed, so it does not keep magnetising gems that drop afterwards. */
  XP.vacuumAll = function () { this.vacuumTimer = 1.5; };

  XP.update = function (dt) {
    const player = WS.Game.player;
    if (this.vacuumTimer > 0) this.vacuumTimer -= dt;
    const vacuum = this.vacuumTimer > 0;

    let i = 0;
    while (i < this.pool.count) {
      const gem = this.pool.active[i];
      const [dx, dy, distance] = WS.normalize(player.x - gem.x, player.y - gem.y);
      if (gem.pop > 0) gem.pop -= dt;
      if (vacuum) {
        gem.x += dx * 900 * dt;
        gem.y += dy * 900 * dt;
      } else if (distance < player.pickupRadius) {
        const pull = WS.max(160, 420 * (1 - distance / player.pickupRadius));
        gem.x += dx * pull * dt;
        gem.y += dy * pull * dt;
      }
      if (distance < player.radius + gem.radius + 5) {
        WS.Game.run.gemsCollected++;
        WS.Save.stats.gemsCollected++;
        if (vacuum) WS.Audio.play('gem');
        const value = gem.value;
        this.pool.releaseAt(i);
        WS.Player.gainXP(player, value);
        if (WS.Game.leveling || !WS.Game.running) return;
      } else i++;
    }
  };

  WS.XP = XP;

})(window.WS);

File: addon/XP.lua
local _, WS = ...

-- Experience gems. Tiered like item qualities (green / blue / purple by
-- value), pooled, and merged into neighbors when the field is saturated so
-- late-game swarms never lose experience to the entity cap.
WS.XP = {}
local XP = WS.XP

local GEM_ICON = "Interface\\Icons\\INV_Misc_Gem_Emerald_02"
local GEM_ICON_MID = "Interface\\Icons\\INV_Misc_Gem_Sapphire_02"
local GEM_ICON_HIGH = "Interface\\Icons\\INV_Misc_Gem_Amethyst_02"

function XP:Initialize()
    self.pool = WS.Pool:New(function()
        local frame = WS.UI:CreateEntityFrame(14)
        WS.UI:ApplyRoundMask(frame) -- gems read as glittering round stones
        return { frame = frame }
    end, function(gem) gem.frame:Hide() end, 100)
    self.vacuumTimer = 0
end

local function Style(gem)
    local colors = WS.Constants.COLORS
    if gem.value >= 25 then
        gem.size = 18
        gem.frame.icon:SetTexture(GEM_ICON_HIGH)
        gem.frame.icon:SetVertexColor(colors.gemHigh[1], colors.gemHigh[2], colors.gemHigh[3])
    elseif gem.value >= 8 then
        gem.size = 15
        gem.frame.icon:SetTexture(GEM_ICON_MID)
        gem.frame.icon:SetVertexColor(colors.gemMid[1], colors.gemMid[2], colors.gemMid[3])
    else
        gem.size = 12
        gem.frame.icon:SetTexture(GEM_ICON)
        gem.frame.icon:SetVertexColor(colors.gemLow[1], colors.gemLow[2], colors.gemLow[3])
    end
end

function XP:SpawnGem(x, y, value)
    if value <= 0 then return end
    -- At the cap, fold the value into an existing gem instead of dropping it.
    if self.pool.count >= WS.Constants.MAX_GEMS then
        local gem = self.pool.active[WS.random(1, self.pool.count)]
        gem.value = gem.value + value
        Style(gem)
        return
    end
    local gem = self.pool:Acquire()
    gem.x, gem.y, gem.value, gem.radius = x, y, value, 8
    Style(gem)
    gem.frame:Show()
    WS.UI:Place(gem.frame, gem.x, gem.y, gem.size)
end

-- Lodestone effect: a brief, one-shot pull that sweeps every gem currently on
-- the field into the survivor. (Timed, so it does not keep magnetizing gems
-- that spawn afterward - that was the "picking up everything without moving"
-- bug.)
function XP:VacuumAll()
    self.vacuumTimer = 1.5
end

function XP:Update(dt)
    local player = WS.Game.player
    if self.vacuumTimer > 0 then self.vacuumTimer = self.vacuumTimer - dt end
    local vacuum = self.vacuumTimer > 0
    local i = 1
    while i <= self.pool.count do
        local gem = self.pool.active[i]
        local dx, dy, distance = WS.Normalize(player.x - gem.x, player.y - gem.y)
        if vacuum then
            gem.x = gem.x + dx * 900 * dt
            gem.y = gem.y + dy * 900 * dt
        elseif distance < player.pickupRadius then
            local pull = WS.max(160, 420 * (1 - distance / player.pickupRadius))
            gem.x = gem.x + dx * pull * dt
            gem.y = gem.y + dy * pull * dt
        end
        if distance < player.radius + gem.radius + 5 then
            WS.Game.run.gemsCollected = WS.Game.run.gemsCollected + 1
            WoWSurvivorsDB.statistics.gemsCollected = WoWSurvivorsDB.statistics.gemsCollected + 1
            -- Silent on ordinary pickup (there are hundreds); only the
            -- lodestone's sweep chimes as gems rush in.
            if vacuum then WS.Audio:Play("gem") end
            local value = gem.value
            self.pool:Release(gem)
            WS.Player:GainXP(player, value)
            if WS.Game.leveling or not WS.Game.running then return end
        else
            WS.UI:Place(gem.frame, gem.x, gem.y, gem.size)
            i = i + 1
        end
    end
end

function XP:Clear()
    self.vacuumTimer = 0
    if self.pool then self.pool:ReleaseAll() end
end
